### DISCUSSION TITLE
How to speedup the warm-up time of ts-node?

### DISCUSSION THREAD

**Discussion by sebastianrath** [TypeStrong/ts-node#1276] [General]
Hello, I am using **ts-node** in the repo [SnowFS](https://github.com/snowtrack/snowfs).

SnowFS is a command-line application written in TypeScript and during unit-tests, I use a batch/shell script to verify the functionality of the program, as sketched below.

![image](https://user-images.githubusercontent.com/12844423/111342823-31bee880-8651-11eb-9fbd-ac812e428211.png)

On Windows: [snowfs.bat](https://github.com/Snowtrack/SnowFS/blob/main/bin/snow#L20)
On macOS and Linux: [snowfs](https://github.com/Snowtrack/SnowFS/blob/main/bin/snow.bat#L22)

As you can see from the sketch above, several tests execute **snow.bat** which uses **ts-node**. Every single call fires up `ts-node`. I found several posts and questions about the warmup time of **ts-node**, but most of them are outdated. What would be the most current way to speedup the warm-up time ot `ts-node`?

As you can see, the unit-tests take very long, for just a few calls (full log [here](https://github.com/Snowtrack/SnowFS/runs/2116757091?check_suite_focus=true#step:7:1))

```

9.cli.ts snow add bar.txt (10s) 
9.cli.ts snow rm foo.txt (25.2s) 
9.cli.ts snow rm subdir (24.9s) 
...
```

Any help is highly appreciated!

**Comment by cspotcode** [comment]
Are you using `transpileOnly` mode?  You can configure `ts-node` options within your tsconfig file: https://github.com/Snowtrack/SnowFS/blob/main/tsconfig.json

The website linked below is still a work-in-progress, but this page explains configuration:
https://typestrong.org/ts-node/docs/configuration

Remember that many people believe ts-node *must* do typechecking, but this is usually not true.  It is actually better to use `tsc` with the `noEmit` option to typecheck *once* before your tests run.  ts-node can skip typechecking.

**Reply by cspotcode** [reply]
I see that we don't have a dedicated Discussion thread for making startup faster.  That will be a good one to create with centralized guidance for improving speed both with and without typechecking enabled.

**Reply by sebastianrath** [reply]
Thank you, that has helped a lot! I used the environment variable `TS_NODE_TRANSPILE_ONLY` inside the shell script / bat.

See: https://github.com/Snowtrack/SnowFS/blob/main/bin/snow.bat#L22
See: https://github.com/Snowtrack/SnowFS/blob/main/bin/snow#L20

The time reduced from almost 10 minutes down to 5!

I also edited the title to make it more generic.

**Reply by cspotcode** [reply]
Re title change: thank you, great idea.

If you are feeling adventurous and want to make it *even faster*, you can try out our experimental swc integration.  swc is a transpiler written in rust.  Our integration lets you take advantage of ts-node's features, automatic tsconfig loading, automatic source-map support, but uses swc to compile even faster.  This feature is unreleased so you will need to install ts-node from our `main` branch.

Check out the "Usage" instructions which have everything you need to try it out: https://github.com/TypeStrong/ts-node/pull/1160

**Reply by sebastianrath** [reply]
That is awesome! I will give this a try!

**Reply by cspotcode** [reply]
Did you have a chance to try this out?  Did it work out well, or were there any problems?

**Reply by sebastianrath** [reply]
I added my changes to my forked repo in [seb-mtl/SnowFS#test-swc-transpiler](https://github.com/seb-mtl/SnowFS/tree/test-swc-transpiler)

The changes are minimal: https://github.com/seb-mtl/SnowFS/commit/5c981baab3f630a5cdbe45aa4be61f7621b81cb6

I basically followed the instruction in #1160 by installing the packages and updated the `tsconfig.json`. Afterwards I executed:

```
cd snowfs.git
./bin/snow.bat drivelist
```

with the error output:

```
internal/modules/cjs/loader.js:414
    throw e;
    ^
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './transspilers/swc-experimental' is not defined by "exports" in
...
```

Quick refresh, the `snow.bat` was the file that executed `ts-node`. I am sure I made a basic mistake here. Any idea what might have happened here?

**Reply by cspotcode** [reply]
Typo in the word transpilers?

**Reply by sebastianrath** [reply]
Ouch, `STRG`+`S` is to blame here, thanks! ;-). It seems to work now, but there is no speed improvement. Given the fact that it complained about `transspilers` before and now doesn't, makes me assume that it does load the `tsconfig.json`. Here are some results:

Without `swc-experimental`:

```
$ time snow.bat driveinfo
real    0m1.407s
user    0m0.000s
sys     0m0.000s
```

With `swc-experimental`:
```
$ time snow.bat driveinfo
real    0m1.397s
user    0m0.000s
sys     0m0.000s
```

**Reply by cspotcode** [reply]
You may also have a typo in the `transpileOnly` option.

**Reply by sebastianrath** [reply]
Yes, I saw that, really weird. I also reset my local repo just to be sure. Please note, with `ts-node` I leave my field of expertise. I might have done something else causing this, still reporting in case it's helpful.

```
.../git/snowfs/node_modules/.bin/ts-node: line 14:  1400 Illegal instruction     node "$basedir/../ts-node/dist/bin.js" "$@"
```

**Edit:** Removing the `"ts-node"` section from the `tsconfig.json` makes it work again.  So it seems to be related to `swc`

**Reply by sebastianrath** [reply]
FYI: I fixed the typos in my GitHub repo.

**Reply by dandv** [reply]
Thanks for that solution, @cspotcode! That brought down my startup time from 5-6 seconds to <1.

I feel this is a *massive* performance improvement and indeed, many users aren't aware of this flag. Might it be worth mentioning it somewhere more prominent? Maybe [here](https://github.com/TypeStrong/ts-node#options-via-tsconfigjson)?

TL;DR - add at the top-level of your `tsconfig.json` this:  `"ts-node": { "transpileOnly": true }`.

Unlike passing CLI flags, this also works when using ESM via `node --loader=ts-node/esm`.

I did however find that the option apparently must be in the closest `tsconfig.json` file. In a monorepo setup, if my `ts-config.json` `"extends": "../tsconfig-base.json"`, and I place the `transpileOnly` option at the top-level of `../tsconfig-base.json"`, the startup is as slow as without it.

**Reply by cspotcode** [reply]
The new website, a work-in-progress, mentions it here: https://typestrong.org/ts-node/docs/configuration

**Reply by cspotcode** [reply]
You're correct about `"extends"` and `ts-node` options: currently they are only pulled from the first tsconfig, not from any extended tsconfigs.  #1292 will add support to use ts-node options from parent tsconfigs.

**Reply by benatshippabo** [reply]
@cspotcode the transpiler option is amazing! Any ETA on when it will be released?

**Reply by cspotcode** [reply]
Great, glad you like it.

I created a milestone to track tickets that need to be finished prior to the release of v10.  It's less work than it looks like.  Mostly the new website's documentation needs to be finished, and it's all markdown, so no coding: just writing and organization.  Good work for new contributors.

#1293

**Reply by dguard** [reply]
@seb-mtl, thank you, man
time reduced from 11 sec to 5 sec

**Reply by cspotcode** [reply]
FYI this has since been simplified greatly.  Check out our `"swc": true` option, which is even faster than transpileOnly.

https://typestrong.org/ts-node/docs/swc

**Reply by dguard** [reply]
@cspotcode 
I am using this executable in nodemon.json

`ts-node -r tsconfig-paths/register src/main.ts`

added in tsconfig.json
```
{
  "ts-node": {
    "swc": true
  },
}
```

my tsconfig.json
```
{
  "ts-node": {
    "swc": true
  },
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es2017",
    "sourceMap": false,
    "outDir": "./dist",
    "baseUrl": "./src",
    "incremental": true,
    "paths": {
      "@components/*": ["components/*"],
      "@endpoints/*": ["endpoints/*"]
    }
  },
  "exclude": ["node_modules", "dist"]
}
```

but it not works, what is wrong?

**Reply by cspotcode** [reply]
That configuration is correct and does work.  There must be something else that is wrong that you did not include in your question.

**Reply by dguard** [reply]
@cspotcode, cool it was also required
`yarn add ts-node`
and it works

4 sec

**Comment by justin-rwazi** [comment]
getting `Error: Cannot find module '@/src/@types/common'` with custom `paths`

create `.swcrc` with paths option but still not working
![image](https://github.com/user-attachments/assets/a8199bbf-89aa-4f8b-8173-5979a8f5fe1f)
![image](https://github.com/user-attachments/assets/7f233e93-e69c-4303-b89e-94d7dddf2ed3)

**Reply by dguard** [reply]
try to use vite and setup alias @, it should work

**Reply by justin-rwazi** [reply]
I'm using Express
is there any way to setup?

**Reply by dguard** [reply]
you may

setup vite independently from scratch, add simple call of express, eg. creating a server and listening
move code from existing base into code executed by vite
